wallet/invoices.c

```diff
@@ -89,13 +92,82 @@ struct invoices *invoices_new(const tal_t *ctx,
 	return invs;
 }
 
+static void install_expiration_timer(struct invoices *invoices,
+				     struct invoice *i,
+				     u64 now);
+static void trigger_expiration(struct invoice *i)
+{
+	struct invoices *invoices = (struct invoices *)tal_parent(i);
+	u64 now = time_now().ts.tv_sec;
+	sqlite3_stmt *stmt;
+	struct invoice_waiter *w;
+	const tal_t *tmpctx;
+
+	/* There may be discrepancies between time_mono
+	 * (used by the timer system) and time_now (used
+	 * by the expiry time measurements). So check that
+	 * time_now is reached. */
+	if (i->expiry_time <= now) {
+		tmpctx = tal_tmpctx(i);
+
+		/* Update in-memory and db. */
+		i->state = EXPIRED;
+		stmt = db_prepare(invoices->db,
+				  "UPDATE invoices"
+				  "   SET state = ?"
+				  " WHERE id = ?;");
+		sqlite3_bind_int(stmt, 1, EXPIRED);
+		sqlite3_bind_int64(stmt, 2, i->id);
+		db_exec_prepared(invoices->db, stmt);
+		/* Wake up all waiters. */
+		while ((w = list_pop(&i->waitone_waiters,
+				     struct invoice_waiter,
+				     list)) != NULL) {
+			tal_steal(tmpctx, w);
+			trigger_invoice_waiter(w, i);
+		}
+
+		tal_free(tmpctx);
+		return;
+	} else
+		install_expiration_timer(invoices, i, now);
+
+}
+static void install_expiration_timer(struct invoices *invoices,
+				     struct invoice *i,
+				     u64 now)
+{
+	struct timerel rel;
+
+	assert(i->state == UNPAID);
+	assert(!i->expiration_timer);
+	assert(i->expiry_time > now);
+
+	rel = time_from_sec(i->expiry_time - now);
```

Comment:
Would drop the now arg, from this function, and use the current time.

Also, drop the expiration_timer pointer, as it's unnecessary in fact.

We can use more precision here:

```
   struct abstime expiry = { .ts = { .tv_sec = i->expiry; } }, now = time_now();

   /* Corner case where expiry is past: 
   if (time_after(expiry, now))
        expiry = now;

   new_reltimer(invoices->timers, i, time_between(expiry, now), trigger_expiration, i);
```

Reply:
Okay!